hub/core/compression/base.py

```diff
@@ -0,0 +1,164 @@
+from typing import Union, Dict
+from abc import ABC, abstractmethod  # type: ignore
+from numcodecs.abc import Codec  # type: ignore
+
+import numpy as np
+from .constants import WEBP_COMPRESSOR_NAME, IMAGE_SHAPE_ERROR_MESSAGE, MSGPACK
+
+
+class BaseNumCodec(ABC):
+    """Base class for numcodec compressors"""
+
+    @abstractmethod
+    def encode(self, input: np.ndarray) -> bytes:
+        pass
+
+    @abstractmethod
+    def decode(self, bytes: bytes) -> Union[np.ndarray, bytes]:
+        pass
+
+    @property
+    @abstractmethod
+    def __name__(self):
+        pass
+
+
+class BaseImgCodec(ABC, Codec):
+    """Base class for image codecs"""
+
+    def __init__(self, single_channel: bool = True) -> None:
+        self.single_channel = single_channel
+
+    @property
+    def __name__(self):
+        raise NotImplementedError()
```

Comment:
you can replace this with `self.__class__.__name__` and remove all subclass `__name__` methods

```python
@property
    def __name__(self):
        return self.__class__.__name__.lower()  # if class name is `PNG`, PNG().__name__ will return "png" as long as it inherits `BaseImgCodec`
```

Full example:
```python
class Foo:
    def __init__(self):
        pass

    @property
    def __name__(self):
        return self.__class__.__name__.lower()

class Bar(Foo):
    def __init__(self):
        pass

print(Foo().__name__)  # prints "foo"
print(Bar().__name__)  # prints "bar"
```